### PR TITLE
Fix: General improvements [ED-15699]

### PR DESCRIPTION
**Pull request by mugurelLupu**
## PR Checklist

- [x] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

**Review by TzviRabinovitch — DISMISSED** [review]

File: core/admin/admin.php [hunk 1]
```diff
@@ -701,6 +701,14 @@ public function admin_action_new_post() {
 
 		$post_data = Utils::get_super_global_value( $_GET, 'post_data' ) ?? [];
 
+		if ( ! current_user_can( 'publish_posts' ) && isset( $post_data['post_status'] ) && $post_data['post_status'] === 'publish' ) {
+			$post_data['post_status'] = 'draft';
+		}
+
+		$allowed_post_data_keys = [ 'post_title', 'post_content', 'post_excerpt', 'post_status', 'post_type' ];
+		$post_data = array_intersect_key( $post_data, array_flip( $allowed_post_data_keys ) );
+		$post_data = array_map( 'sanitize_text_field', $post_data );
```
Comment (mykytamurzin) [inline core/admin/admin.php:710]: 708 - 710 doesn't look clear - let's move to a separate function with self-descriptive name. Also let's add a unit test.

File: core/admin/admin.php [hunk 2]
```diff
@@ -731,6 +737,12 @@ public function admin_action_new_post() {
 		die;
 	}
 
+	private function sanitaze_post_data( $post_data ) {
+		$allowed_post_data_keys = [ 'post_title', 'post_content', 'post_excerpt', 'post_status', 'post_type' ];
+		$post_data = array_intersect_key( $post_data, array_flip( $allowed_post_data_keys ) );
+		$post_data = array_map( 'sanitize_text_field', $post_data );
+
+	}
```
Comment (louiswol94) [inline core/admin/admin.php:745]:
```suggestion
		$post_data = array_map( 'sanitize_text_field', $post_data );
	}
```

Comment (louiswol94) [reply core/admin/admin.php:745]: Should we sanitize the individual fields? Something like:

```
$post_data = [
        'post_title' => isset( $post_data['post_title'] ) ? sanitize_text_field( $post_data['post_title'] ) : '',
        'post_content' => isset( $post_data['post_content'] ) ? sanitize_textarea_field( $post_data['post_content'] ) : '',
        'post_status' => isset( $post_data['post_status'] ) ? $post_data['post_status'] : 'draft',
        'post_type' => $post_type,
    ];
```

File: core/admin/admin.php [hunk 3]
```diff
@@ -731,6 +737,13 @@ public function admin_action_new_post() {
 		die;
 	}
 
+	private function sanitaze_post_data( $post_data ) {
```
Comment (louiswol94) [inline core/admin/admin.php:740]:
```suggestion
	private function sanitaze_post_data( $post_data ) {
```

File: core/admin/admin.php [hunk 4]
```diff
@@ -701,6 +701,12 @@ public function admin_action_new_post() {
 
 		$post_data = Utils::get_super_global_value( $_GET, 'post_data' ) ?? [];
 
+		if ( ! current_user_can( 'publish_posts' ) && isset( $post_data['post_status'] ) && 'publish' === $post_data['post_status'] ) {
+			$post_data['post_status'] = 'draft';
+		}
+
+		$post_data = $this->sanitaze_post_data( $post_data );
```
Comment (louiswol94) [inline core/admin/admin.php:708]:
```suggestion
		$post_data = $this->sanitize_post_data( $post_data );
```

File: core/admin/admin.php (hunk 3 above)
Comment (louiswol94) [inline core/admin/admin.php:740]:
```suggestion
	private function sanitize_post_data( $post_data ) {
```

File: core/admin/admin.php [hunk 5]
```diff
@@ -731,6 +737,14 @@ public function admin_action_new_post() {
 		die;
 	}
 
+	private function sanitize_post_data( $post_data ) {
+		$post_data = [
+			'post_title' => isset( $post_data['post_title'] ) ? sanitize_text_field( $post_data['post_title'] ) : '',
+			'post_content' => isset( $post_data['post_content'] ) ? sanitize_textarea_field( $post_data['post_content'] ) : '',
+			'post_status' => isset( $post_data['post_status'] ) ? $post_data['post_status'] : 'draft',
+		];
```
Comment (louiswol94) [inline core/admin/admin.php:745]: So can we only save these 3 keys? What about post_type, post_excerpt etc?

Comment (mugurelLupu) [reply core/admin/admin.php:745]: Post type is added later in the main method, i am not sure about post_excerpt exactly

Comment (louiswol94) [reply core/admin/admin.php:745]: Let's get an allowed list of fields per role, and just remove any fields that are not in the whitelist.

File: core/admin/admin.php [hunk 6]
```diff
@@ -701,6 +701,12 @@ public function admin_action_new_post() {
 
 		$post_data = Utils::get_super_global_value( $_GET, 'post_data' ) ?? [];
 
+		if ( ! current_user_can( 'publish_posts' ) && isset( $post_data['post_status'] ) && 'publish' === $post_data['post_status'] ) {
+			$post_data['post_status'] = 'draft';
```
Comment (louiswol94) [inline core/admin/admin.php:705]: @mugurelLupu What is the purpose of setting a `post_status` here, if you will just remove again it below? `wp_insert_post()` will make it  draft automatically if post_status is not specified. Can you please double check this aspect? Thanks!

**Review by louiswol94 — CHANGES_REQUESTED** [review]

**Review by louiswol94 — APPROVED** [review]

**Pull request merged**